src/main/java/com/atlassian/db/replica/api/DualConnection.java

```diff
@@ -520,14 +555,28 @@ private Builder(
             return this;
         }
 
+        /**
+         * Puts this connection in compatibility mode with a previous version. Developers can use this method to
+         * roll out the new version of the library with a feature flag.
+         *
+         * It's best-effort, and there's no guarantee the library in compatibility mode will always behave
+         * the same way as the previous version of the library.
```

Comment:
> here's no guarantee the library in compatibility mode will always behave
>          * the same way as the previous version of the library

Why? Because still there is room for bugs?
I guess all `compatibleWithPreviousVersion` changes have to be covered in unit tests.
I'm missing example.

What we can do to guarantee compatibility mode working the same way as the previous version ? 
Maybe we should provide 'template' with supplier before if, after if, AKA rolloutService?

Reply:
> What we can do to guarantee compatibility mode working the same way as the previous version ?

I guess it's not possible to give such a guarantee. It's better to say it explicit than to pretend we can make backwards-compatible every change.